### DISCUSSION TITLE
Exposed Registry

### DISCUSSION THREAD

**Discussion by darren-oxford** [okd-project/okd#1301] [General]
Having recently updated to the OKD 4.11 release I'm now unable to login to the internal registry from outside the cluster with podman.

I can login within the cluster 'podman login -u myusername -p $(oc whoami -t) image-registry.openshift-image-registry.svc:5000' so I know that the registry is running, authentication is working and the service is OK.

The registry is exposed using the usual spec:defaultRoute:true and I can verify it by curling the url and get a response {"access_token":"anonymous","token":"anonymous"}, this confirms the route is working. However when attempting to login, I can see in the registry logs that I am getting a 401.

Seems very odd that it is working internally but not externally, it is almost as though there is some IP based access control?

Anyone have any suggestions?

**Comment by vrutkovs** [comment]
Does `oc login ... && oc registry login`` work? That's what I usually do to pass token to podman

**Reply by darren-oxford** [reply]
Thanks for the suggestion @vrutkovs, unfortunately, while oc registry login does save the token and podman finds it ok, I still get a 401.

`http.request.uri=/v2/ http.request.useragent=Go-http-client/1.1 http.response.contenttype="application/json; charset=utf-8" http.response.duration=2.235907ms http.response.status=401 http.response.written=87`

**Comment by darren-oxford** [comment]
After a lot of digging, I have found the what I believe to be the cause of my problem.

I have a bare metal cluster with HAProxy, I run HAProxy in a container with the send-proxy-v2 option and in order to run it rootless have ports exposed as -p 80:8080 -p 443:4443. HAProxy frontend bound to port 4443 and making requests to 443 on the backend. The problem with this is that somewhere along the line podman is seeing the 4443 port somewhere along the line.

It must be HAProxy that is passing the port information to the route and then that is passed to the registry service. Because I am using tcp mode (level 4) I am unable to set X-Forwarded-Port on HAProxy. An odd one, but thought I'd report back as it seems the cause is this https://docs.docker.com/registry/deploying/#importantrequired-http-headers
Why it has changed between OKD 4.10 and 4.11 I don't know.

This is obviously a peculiar edge case and I will probably have to rethink the set-up, unless there is a way to force HAProxy to report the correct port.
